components/raftstore-v2/src/operation/command/admin/compact_log.rs

```diff
@@ -0,0 +1,279 @@
+        req: RaftCmdRequest,
+    ) -> Result<u64> {
+        let compact_log = req.get_admin_request().get_compact_log();
+        // TODO: add unit tests to cover all the message integrity checks.
+        if compact_log.get_compact_term() == 0 {
+            info!(
+                self.logger,
+                "compact term missing, skip";
+                "command" => ?compact_log
+            );
+            // old format compact log command, safe to ignore.
+            return Err(box_err!(
+                "command format is outdated, please upgrade leader"
+            ));
+        }
+
+        let data = req.write_to_bytes().unwrap();
+        self.propose_with_ctx(store_ctx, data, vec![])
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    pub fn apply_compact_log(
+        &mut self,
+        req: &AdminRequest,
+        _log_index: u64,
+    ) -> Result<(AdminResponse, AdminCmdResult)> {
+        Ok((
+            AdminResponse::default(),
+            AdminCmdResult::CompactLog(CompactLogResult {
+                compact_index: req.get_compact_log().get_compact_index(),
+                compact_term: req.get_compact_log().get_compact_term(),
+            }),
+        ))
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn on_apply_res_compact_log<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        res: CompactLogResult,
+    ) {
+        let first_index = self.entry_storage().first_index();
+        if res.compact_index <= first_index {
+            debug!(
+                self.logger,
+                "compact index <= first index, no need to compact";
+                "compact_index" => res.compact_index,
+                "first_index" => first_index,
+            );
+            return;
+        }
+        // TODO: check is_merging
+        // compact failure is safe to be omitted, no need to assert.
+        if res.compact_index <= self.entry_storage().truncated_index() {
+            return;
+        }
+
+        // TODO: check entry_cache_warmup_state
+        self.entry_storage_mut()
+            .compact_entry_cache(res.compact_index);
+        self.storage_mut()
+            .cancel_generating_snap(Some(res.compact_index));
+
+        self.entry_storage_mut()
+            .apply_state_mut()
+            .mut_truncated_state()
+            .set_index(res.compact_index);
+        self.entry_storage_mut()
+            .apply_state_mut()
+            .mut_truncated_state()
+            .set_term(res.compact_term);
+
+        self.maybe_compact_log_from_engine(store_ctx);
+    }
+
+    pub fn maybe_compact_log_from_engine<T>(&mut self, store_ctx: &mut StoreContext<EK, ER, T>) {
+        let compact_index = std::cmp::min(
+            self.entry_storage().truncated_index(),
+            self.storage().apply_trace().persisted_apply_index(),
+        );
+        if compact_index > self.last_engine_compact_log_index() {
```

Comment:
`last_engine_compact_log_index` may not be necessary. For `on_ready_res`, it should compact log if `truncated_index < persisted_applied` before applying the updates, For `record_apply_trace`, it should compact log if `persisted_applied < truncated_index`.

Reply:
You are right. But it makes the code a little more complex, and it introduces an extra dependency between caller and callee.

Reply:
On a second thought, I'll just add a new function `on_persisted_apply_advanced`, which is probably useful for tablet gc too.

Reply:
Why? For example, `truncated_index` is advanced from 3 to 10 in `on_ready_res`, while `persisted_applied is 5. Because 3 < 5, so it should gc logs using `cmp::min(10, 5) = 5`.

Reply:
I see. It is still very complex.

Reply:
It's just two if check to get rid of a field (and two if check).